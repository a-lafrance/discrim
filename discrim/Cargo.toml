[package]
name = "discrim"
version = "0.1.1"
edition = "2021"

authors = ["Arthur Lafrance <lafranca@uci.edu>"]
description = "Construct enum variants from their tag"
license = "MIT"
readme = "../README.md"
repository = "https://github.com/a-lafrance/discrim"
keywords = ["enum", "discriminant", "derive", "codegen"]

[dependencies]
discrim-codegen = { version = "0.1.1", path = "../discrim-codegen" }
