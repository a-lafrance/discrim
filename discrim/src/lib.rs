pub use discrim_codegen::*;

// TODO: better error handling, specifically maybe a more robust type than just storing the discriminant
// Should never be manually implemented
pub trait FromDiscriminant<T>: Sized {
    fn from_discriminant(tag: T) -> Result<Self, T>;
}
