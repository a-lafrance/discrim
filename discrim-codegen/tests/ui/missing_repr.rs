use discrim::FromDiscriminant;

#[derive(FromDiscriminant)]
enum MissingRepr {
    Bad,
}

fn main() {
    println!("missing repr");
}
