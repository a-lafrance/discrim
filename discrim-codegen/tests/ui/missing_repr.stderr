error: proc-macro derive panicked
 --> $DIR/missing_repr.rs:3:10
  |
3 | #[derive(FromDiscriminant)]
  |          ^^^^^^^^^^^^^^^^
  |
  = help: message: #[repr(...)] attribute is required
